qiskit/providers/backend.py

```diff
@@ -229,3 +241,443 @@ class can handle either situation.
+    exposing equivalent required immutable properties of the backend device
+    are added. For example ``backend.configuration().n_qubits`` is accessible
+    from ``backend.num_qubits`` now.
+
+    The ``options`` attribute of the backend is used to contain the dynamic
+    user configurable options of the backend. It should be used more for
+    runtime options that configure how the backend is used. For example,
+    something like a ``shots`` field for a backend that runs experiments which
+    would contain an int for how many shots to execute.
+
+    If migrating a provider from :class:`~qiskit.providers.BackendV1` or
+    :class:`~qiskit.providers.BaseBackend` one thing to keep in mind is for
+    backwards compatibility you might need to add a configuration method that
+    will build a :class:`~qiskit.providers.models.BackendConfiguration` object
+    and :class:`~qiskit.providers.models.BackendProperties` from the attributes
+    defined in this class for backwards compatibility.
+    """
+
+    version = 2
+
+    def __init__(
+        self,
+        provider: Provider = None,
+        name: str = None,
+        description: str = None,
+        online_date: datetime.datetime = None,
+        backend_version: str = None,
+        **fields,
+    ):
+        """Initialize a BackendV2 based backend
+
+        Args:
+            provider: An optional backwards reference to the
+                :class:`~qiskit.providers.Provider` object that the backend
+                is from
+            name: An optional name for the backend
+            description: An optional description of the backend
+            online_date: An optional datetime the backend was brought online
+            backend_version: An optional backend version string. This differs
+                from the :attr:`~qiskit.providers.BackendV2.version` attribute
+                as :attr:`~qiskit.providers.BackendV2.version` is for the
+                abstract :attr:`~qiskit.providers.Backend` abstract interface
+                version of the object while ``backend_version`` is for
+                versioning the backend itself.
+
+        Raises:
+            AttributeError: If a field is specified that's outside the backend's
+                options
+        """
+
+        self._options = self._default_options()
+        self._provider = provider
+        self._coupling_map = None
+        if fields:
+            for field in fields:
+                if field not in self._options.data:
+                    raise AttributeError("Options field %s is not valid for this backend" % field)
+            self._options.update_config(**fields)
+        self._basis_gates_all = None
+        self.name = name
+        self.description = description
+        self.online_date = online_date
+        self.backend_version = backend_version
+
+    @property
+    def instructions(self) -> List[Tuple[Instruction, Tuple[int]]]:
```

Comment:
I actually think we might want this to return classical functions too if they're in a backend's target. It would just be an entry like `(ForLoop, ())`.  For the classical functions at least the classical bits don't matter as much because they should be usable on all of them (at least in the qiskit model). But, I'll defer to what @ajavadia would like here since he asked for this method.

As for having a custom class return we're looking at creating a Qiskit wide class for this in: https://github.com/Qiskit/qiskit-terra/pull/7020 (which will replace the entries in `QuantumCircuit.data`) but it's a bit stalled right now because the name people want to use for this class is `Instruction` which is possibly the most complicated choice because of the two already existing `Instruction` classes in terra with `circuit.Instruction` hopefully eventually being transitioned to `Operation`. But that's a much longer road than this PR. I was thinking since that new Instruction class will likely be a `NamedTuple` we can probably reuse it here but thinking about it again you're right we don't want a mismatch in the length of the argument and probably should add a custom class here too (although I don't like the character count on `PhysicalQubitInstruction` :) ).

Reply:
Yeah part of the reason I asked for the name `Operation` in https://github.com/Qiskit/qiskit-terra/pull/7087, not `QuantumOperation`, was that it could represent classical operations too. At least in the qiskit circuit model Operations consume either quantum bits, or classical bits, or both. So they can be represented by nodes in the DAG. A classical function that reads some bits and writes to some bits would block both the read and write resources. 

With the introduction of types we may want to change this. I guess the example Thomas has in mind is if a backend provides a cosine() function which converts between two fixed-point classical numbers. Fundamentally I think this can be done with the bit model but building types on it makes life easier of course. I think this requires a larger discussion though.

Reply:
Regarding #7020 why not just choose something like `BaseInstruction` or something like that to disambiguate? My major concern is avoiding coupling this class to an inflexible representation based on primitive tuples rather than a class we can migrate in the future.

Reply:
This may be a comment for #7020 but I think it has `__slots__` for speed. Though I agree there may be some classes that need other things. An example would be good.